src/utils/label.ts

```diff
@@ -87,20 +90,41 @@ function hasAncestorCustomElements(label: HTMLCalciteLabelElement, componentEl:
 export function connectLabel(component: LabelableComponent): void {
   const labelEl = findLabelForComponent(component.el);
 
-  if (!labelEl || onLabelClickMap.has(labelEl)) {
+  const addClickEventListenerToComponentLabel = () => {
+    component.labelEl = labelEl;
+    const boundOnLabelClick = onLabelClick.bind(component);
+    onLabelClickMap.set(component.labelEl, boundOnLabelClick);
+    component.labelEl.addEventListener(labelClickEvent, boundOnLabelClick);
+  };
+
+  const boundOnLabelConnected = onLabelConnected.bind(component);
+  const boundOnLabelDisconnected = onLabelDisconnected.bind(component);
+
+  if (onLabelClickMap.has(labelEl)) {
     return;
   }
 
-  component.labelEl = labelEl;
-  const boundOnLabelClick = onLabelClick.bind(component);
-  onLabelClickMap.set(component.labelEl, boundOnLabelClick);
-  component.labelEl.addEventListener(labelClickEvent, boundOnLabelClick);
+  if (labelEl) {
+    addClickEventListenerToComponentLabel();
+    unlabeledComponents.delete(component);
+    document.removeEventListener(labelConnectedEvent, boundOnLabelConnected);
+    document.addEventListener(labelDisconnectedEvent, boundOnLabelDisconnected);
+  } else {
+    boundOnLabelDisconnected();
```

Comment:
Does this need to be skipped if the labelable is already in `unlabeledComponents`? Otherwise, I think you may end up with multiple listeners in the following scenario:

```html
<calcite-some-labelable id="foo"></calcite-some-labelable> 
<!-- label not found, listens for connected event -->

<calcite-some-labelable id="bar"></calcite-some-labelable> 
<!-- label not found, listens for connected event -->
```

<img src=https://user-images.githubusercontent.com/197440/154555856-a4f2596b-1e92-4a6a-80b3-9066dc72c0bb.jpg width=200>

```html
<calcite-label for="foo">foo</calcite-label> 
<!-- added! fires connected event -->

<calcite-some-labelable id="foo"></calcite-some-labelable>
<!-- label found, gets wired up, cleans up listeners -->

<calcite-some-labelable id="bar"></calcite-some-labelable>
<!-- label not found, listens for connected event **again** -->
```